docs/design-proposals/tidb-stable-scheduling.md

```diff
@@ -0,0 +1,167 @@
+write a new predicate function for pods of TiDB server. In this new predicate
+function we can choose the its previous node of TiDB member it it exists in candidate
+nodes.
+
+## Implementation
+
+At first, we track assigned node of TiDB member in status of TiDBCluster.
+
+```
+type TiDBMember struct {
+  ...
+	// Node hosting pod of this TiDB member.
+	NodeName string `json:"node,omitempty"`
+}
+```
+
+In new predicate `StableScheduling`, we filter out other nodes for TiDB pod if
+previous node for this TiDB member exists in candidate nodes.
+
+## Alternatives
+
+### Deploy TiDB members on all nodes
+
+If TiDB members are running on all nodes, clients can access TiDB server with
+any node IP address.
+
+We can achieve this by using DaemonSet or Deployment with pod anti-affinity.
+But if the Kubernetes cluster is large, to deploy TiDB members of each TiDB
+cluster on every node is very inefficient and will consume too many resources.
+
+### Leverage load balancer health check mechanism
+
+Load balancers often have health checks on its backends, it will remove invalid
+backend automatically. We can add all nodes into load balancer. Drawbacks with
+this solution are:
+
+- maybe too heavy for LB if the Kubernetes cluster is large
+  - `NumberOfTiDBClusters` x `NumberOfNodess ports should be health checked
+- need to add every new node into the backend of LB
+- hard to monitor LB (not all failed backends must be fixed)
+
+Some of them can be alleviated by restricting TiDB members in a fixed set of
+nodes (by using NodeSelector/NodeSelector/Taints&Tolerations). This requires
+the user to pre-select the nodes to run TiDB pods.
+
+### Use local PV to bind TiDB member to a nod
+
+Local PVs are local resources of a node. If a TiDB member is using a local PV
+on a node, this node is the only available node for it.
+
+This solution is like using a fixed set of nodes to run TiDB pods and does not
+require the user to pre-select nodes. But every pod is bound to one node, its
+new pod will be pending forever if resources of the node are consumed by other
+nodes.
+
+Another drawback is it requires to bind a dummy PV for each TiDB pod which is
+complex to manage and will confuse the user.
+
+## Limitations
+
+### No guarantee if there is another scheduler to schedule pods to TiDB nodes
+
+This is because when old pod of TiDB member is terminated, resources of its
+node are released and other schedulers may schedule other pods onto this node.
+When `tidb-scheduler` schedules new pod of TiDB member, its previous node may
+not fit.
+
+#### Workarounds
+
+- Reserve nodes for TiDB members (e.g. by taints)
+
+### tidb-cluster cannot schedule new pod of TiDB member back to its node if it does not meet new requirements
+
+If we upgrade TiDB pods to request more resources, it is possible that its node node
+may not have enough resources for the new pod.
+
+It applies if some other scheduling requirements are changed, e.g.
```

Comment:
This should be written as a warning in the document.

Reply:
I added a note about this and workflow in the Proposal section. All other comments are addressed by the way.